### PR TITLE
[BOLT][DWARF][NFC] Refactor address writers

### PR DESCRIPTION
**Pull request by sayhaan**
Refactors address writers to create an instance for each CU and its DWO CU.

### CODE REVIEW

File: bolt/include/bolt/Core/DebugData.h [hunk 1]
```diff
@@ -344,14 +344,34 @@ class DebugAddrWriter {
...
+                              const uint64_t Offset);
+
+  /// Appends an AddressSectionBuffer to the address writer buffer for the given
+  /// CU.
+  void appendToAddressBuffer(const AddressSectionBuffer &Buffer) {
+    *AddressStream << Buffer;
+  }
+
+  /// Sets AddressByteSize for the CU.
+  void setAddressByteSize(uint8_t AddressByteSize) {
```
Comment (ayermolo) [inline bolt/include/bolt/Core/DebugData.h:367]: const

File: bolt/include/bolt/Core/DebugData.h [hunk 2]
```diff
@@ -396,6 +416,8 @@ class DebugAddrWriter {
 
     void dump();
 
+    bool empty = false;
```
Comment (ayermolo) [inline bolt/include/bolt/Core/DebugData.h:419]: Shouldn't default be true?

File: bolt/lib/Rewrite/DWARFRewriter.cpp [hunk 3]
```diff
@@ -615,25 +615,30 @@ void DWARFRewriter::updateDebugInfo() {
...
   std::mutex AccessMutex;
   // Needs to be invoked in the same order as CUs are processed.
   auto createRangeLocList = [&](DWARFUnit &CU) -> DebugLocWriter * {
     std::lock_guard<std::mutex> Lock(AccessMutex);
     const uint16_t DwarfVersion = CU.getVersion();
     if (DwarfVersion >= 5) {
-      LocListWritersByCU[CUIndex] =
-          std::make_unique<DebugLoclistWriter>(CU, DwarfVersion, false);
+      auto AddrW = std::make_unique<DebugAddrWriterDwarf5>(&BC);
+      AddressWritersByCU[CU.getOffset()] = std::move(AddrW);
```
Comment (ayermolo) [inline bolt/lib/Rewrite/DWARFRewriter.cpp:633]: Look into insert so you don't have to access unordered map multiple times.

File: bolt/lib/Rewrite/DWARFRewriter.cpp [hunk 4]
```diff
@@ -645,6 +650,10 @@ void DWARFRewriter::updateDebugInfo() {
       }
 
     } else {
+      auto AddrW = std::make_unique<DebugAddrWriter>(&BC);
+      AddressWritersByCU[CU.getOffset()] = std::move(AddrW);
```
Comment (ayermolo) [inline bolt/lib/Rewrite/DWARFRewriter.cpp:654]: insert

File: bolt/lib/Rewrite/DWARFRewriter.cpp [hunk 5]
```diff
@@ -677,6 +686,7 @@ void DWARFRewriter::updateDebugInfo() {
     DebugRangesSectionWriter *RangesSectionWriter =
         Unit->getVersion() >= 5 ? RangeListsSectionWriter.get()
                                 : LegacyRangesSectionWriter.get();
+    auto &AddressWriter = AddressWritersByCU[Unit->getOffset()];
```
Comment (ayermolo) [inline bolt/lib/Rewrite/DWARFRewriter.cpp:689]: don't use auto

File: bolt/lib/Rewrite/DWARFRewriter.cpp [hunk 6]
```diff
@@ -1567,14 +1577,10 @@ void DWARFRewriter::finalizeDebugSections(
...
 
-  // AddrWriter should be finalized after debug_loc since more addresses can be
-  // added there.
-  if (AddrWriter->isInitialized()) {
-    AddressSectionBuffer AddressSectionContents = AddrWriter->finalize();
-    BC.registerOrUpdateNoteSection(".debug_addr",
-                                   copyByteArray(AddressSectionContents),
-                                   AddressSectionContents.size());
-  }
+  AddressSectionBuffer AddressSectionContents = AddrWriter->finalize();
```
Comment (ayermolo) [inline bolt/lib/Rewrite/DWARFRewriter.cpp:1580]: Should still check if it's initialized. 
Also noticed small "bug" in original code. Should store in reference, and make const

File: bolt/lib/Rewrite/DWARFRewriter.cpp [hunk 7]
```diff
@@ -1628,6 +1634,16 @@ void DWARFRewriter::finalizeCompileUnits(DIEBuilder &DIEBlder,
                                          CUOffsetMap &CUMap,
                                          const std::list<DWARFUnit *> &CUs) {
   for (DWARFUnit *CU : CUs) {
+    auto AddressWriterIterator = AddressWritersByCU.find(CU->getOffset());
+    assert(AddressWriterIterator != AddressWritersByCU.end() &&
+           "AddressWriter does not exist for CU");
+    auto &AddressWriter = AddressWriterIterator->second;
```
Comment (ayermolo) [inline bolt/lib/Rewrite/DWARFRewriter.cpp:1640]: don't use auto

File: bolt/lib/Rewrite/DWARFRewriter.cpp [hunk 8]
```diff
@@ -1628,6 +1634,16 @@ void DWARFRewriter::finalizeCompileUnits(DIEBuilder &DIEBlder,
...
+    AddressWriter->update();
```
Comment (ayermolo) [inline bolt/lib/Rewrite/DWARFRewriter.cpp:1641]: Name no longer fits quite right. Maybe change to finalize, and finalize to like releaseBuffer() with the same semantic as for DebugRangeListsSectionWriter.

File: bolt/lib/Rewrite/DWARFRewriter.cpp [hunk 9]
```diff
@@ -1628,6 +1634,16 @@ void DWARFRewriter::finalizeCompileUnits(DIEBuilder &DIEBlder,
...
+    AddressWriter->updateAddrBase(DIEBlder, *CU, AddrWriter->finalize().size());
+    if (AddressWriter->isInitialized()) {
```
Comment (ayermolo) [inline bolt/lib/Rewrite/DWARFRewriter.cpp:1643]: shouldn't this check also capture updateAddrBase?

File: bolt/lib/Rewrite/DWARFRewriter.cpp [hunk 10]
```diff
@@ -1628,6 +1634,16 @@ void DWARFRewriter::finalizeCompileUnits(DIEBuilder &DIEBlder,
...
+      AddressSectionBuffer AddressSectionContents = AddressWriter->finalize();
+      AddrWriter->appendToAddressBuffer(AddressSectionContents);
```
Comment (ayermolo) [inline bolt/lib/Rewrite/DWARFRewriter.cpp:1645]: Maybe rename AddrWriter to something like FinalAddrWriter or CombinedAddrWriter.

File: bolt/lib/Core/DebugData.cpp [hunk 11]
```diff
@@ -393,6 +393,7 @@ void DebugARangesSectionWriter::writeARangesSection(
 DebugAddrWriter::DebugAddrWriter(BinaryContext *BC) : BC(BC) {
   Buffer = std::make_unique<AddressSectionBuffer>();
   AddressStream = std::make_unique<raw_svector_ostream>(*Buffer);
+  Map = AddressForDWOCU();
```
Comment (ayermolo) [inline bolt/lib/Core/DebugData.cpp:396]: I don't believe this is necessary. It's a member of your class. Default constructor will already be invoked when class instantiated.

File: bolt/lib/Core/DebugData.cpp [hunk 12]
```diff
@@ -405,11 +406,8 @@ void DebugAddrWriter::AddressForDWOCU::dump() {
 }
 uint32_t DebugAddrWriter::getIndexFromAddress(uint64_t Address, DWARFUnit &CU) {
   std::lock_guard<std::mutex> Lock(WriterMutex);
-  const uint64_t CUID = getCUID(CU);
-  if (!AddressMaps.count(CUID))
-    AddressMaps[CUID] = AddressForDWOCU();
-
-  AddressForDWOCU &Map = AddressMaps[CUID];
+  if (Map.begin() == Map.end())
+    Map.empty = true;
```
Comment (ayermolo) [inline bolt/lib/Core/DebugData.cpp:410]: This whole Map.empty logic seems reversed to me. Once getIndexFromAddress is called, it should be false.
Also make it private and add setter/getter.

File: bolt/lib/Core/DebugData.cpp [hunk 13]
```diff
@@ -449,25 +447,20 @@ static void updateAddressBase(DIEBuilder &DIEBlder, DebugAddrWriter &AddrWriter,
...
     return;
-  std::vector<IndexAddressPair> SortedMap(AM->second.indexToAddressBegin(),
-                                          AM->second.indexToAdddessEnd());
+  std::vector<IndexAddressPair> SortedMap(Map.indexToAddressBegin(),
+                                          Map.indexToAdddessEnd());
   // Sorting address in increasing order of indices.
   llvm::sort(SortedMap, llvm::less_first());
 
-  uint8_t AddrSize = CU.getAddressByteSize();
+  uint8_t AddrSize = AddressByteSize;
```
Comment (ayermolo) [inline bolt/lib/Core/DebugData.cpp:463]: Why not just use AddressByteSize directly?

File: bolt/include/bolt/Core/DebugData.h [hunk 14]
```diff
@@ -408,14 +430,16 @@ class DebugAddrWriter {
...
+  AddressForDWOCU Map;
+  uint8_t AddressByteSize;
+  std::optional<uint64_t> AddrOffsetSectionBase;
   /// Mutex used for parallel processing of debug info.
   std::mutex WriterMutex;
   std::unique_ptr<AddressSectionBuffer> Buffer;
   std::unique_ptr<raw_svector_ostream> AddressStream;
   /// Used to track sections that were not modified so that they can be re-used.
-  DenseMap<uint64_t, uint64_t> UnmodifiedAddressOffsets;
+  static DenseMap<uint64_t, uint64_t> UnmodifiedAddressOffsets;
```
Comment (ayermolo) [inline bolt/include/bolt/Core/DebugData.h:442]: For now it's fine, but for multi threaded step you will need to think of locks since this is shared between instances.

File: bolt/lib/Core/DebugData.cpp [hunk 15]
```diff
@@ -449,25 +447,20 @@ static void updateAddressBase(DIEBuilder &DIEBlder, DebugAddrWriter &AddrWriter,
...
-  uint64_t Offset = Buffer->size();
-  // If does not exist this CUs DWO section didn't contribute to .debug_addr.
-  if (AM == AddressMaps.end())
+void DebugAddrWriter::updateAddrBase(DIEBuilder &DIEBlder, DWARFUnit &CU,
+                                     const uint64_t Offset) {
+  updateAddressBase(DIEBlder, *this, CU, Offset);
+}
+
+void DebugAddrWriter::update() {
+  if (Map.indexToAddressBegin() == Map.indexToAdddessEnd())
```
Comment (ayermolo) [inline bolt/lib/Core/DebugData.cpp:456]: You added isInitialized() why not use that for this and other instances where you do this check.
Overall I don't quite understand the point of empty bool you have added. Doesn't seem necessary.

File: bolt/lib/Core/DebugData.cpp [hunk 16]
```diff
@@ -505,22 +511,18 @@ void DebugAddrWriterDwarf5::update(DIEBuilder &DIEBlder, DWARFUnit &CU) {
   DWARFDebugAddrTable AddrTable;
   DIDumpOptions DumpOpts;
   constexpr uint32_t HeaderSize = 8;
-  const uint64_t CUID = getCUID(CU);
-  const uint8_t AddrSize = CU.getAddressByteSize();
-  auto AMIter = AddressMaps.find(CUID);
+  const uint8_t AddrSize = AddressByteSize;
```
Comment (ayermolo) [inline bolt/lib/Core/DebugData.cpp:514]: why not use AddressByteSize directly?

File: bolt/lib/Core/DebugData.cpp [hunk 17]
```diff
@@ -505,22 +511,18 @@ void DebugAddrWriterDwarf5::update(DIEBuilder &DIEBlder, DWARFUnit &CU) {
   DWARFDebugAddrTable AddrTable;
   DIDumpOptions DumpOpts;
   constexpr uint32_t HeaderSize = 8;
```
Comment (ayermolo) [inline bolt/lib/Core/DebugData.cpp:513]: Extract this to class level so you can also use it in DebugAddrWriterDwarf5::updateAddrBase instead of doing + 8.

File: bolt/lib/Rewrite/DWARFRewriter.cpp [hunk 18]
```diff
@@ -1628,6 +1640,24 @@ void DWARFRewriter::finalizeCompileUnits(DIEBuilder &DIEBlder,
                                          CUOffsetMap &CUMap,
                                          const std::list<DWARFUnit *> &CUs) {
   for (DWARFUnit *CU : CUs) {
+    auto AddressWriterIterator = AddressWritersByCU.find(CU->getOffset());
+    assert(AddressWriterIterator != AddressWritersByCU.end() &&
+           "AddressWriter does not exist for CU");
+    std::unique_ptr<DebugAddrWriter> &AddressWriter =
```
Comment (ayermolo) [inline bolt/lib/Rewrite/DWARFRewriter.cpp:1646]: work with underlying instance.

File: bolt/lib/Rewrite/DWARFRewriter.cpp [hunk 19]
```diff
@@ -613,27 +613,33 @@ void DWARFRewriter::updateDebugInfo() {
...
 
   if (BC.isDWARFLegacyUsed())
     LegacyRangesSectionWriter = std::make_unique<DebugRangesSectionWriter>();
 
-  DebugLoclistWriter::setAddressWriter(AddrWriter.get());
-
   uint32_t CUIndex = 0;
   std::mutex AccessMutex;
   // Needs to be invoked in the same order as CUs are processed.
   auto createRangeLocList = [&](DWARFUnit &CU) -> DebugLocWriter * {
```
Comment (ayermolo) [inline bolt/lib/Rewrite/DWARFRewriter.cpp:628]: Update what this is called to reflect that it creates address writers now.

File: bolt/lib/Core/DebugData.cpp [hunk 20]
```diff
@@ -490,10 +479,19 @@ void DebugAddrWriter::update(DIEBuilder &DIEBlder, DWARFUnit &CU) {
...
 }
 
-void DebugAddrWriterDwarf5::update(DIEBuilder &DIEBlder, DWARFUnit &CU) {
+void DebugAddrWriterDwarf5::updateAddrBase(DIEBuilder &DIEBlder, DWARFUnit &CU,
+                                           const uint64_t Offset) {
+  /// Header for DWARF5 has size 8, so we add it to the offset.
+  updateAddressBase(DIEBlder, *this, CU, Offset + HeaderSize);
+}
+
+DenseMap<uint64_t, uint64_t> DebugAddrWriter::UnmodifiedAddressOffsets;
```
[the remark this answers is not in the document]
Comment (sayhaan) [reply bolt/lib/Core/DebugData.cpp:491]: This is the static one, wouldn't we need to define it in the .cpp file since it's static?

https://stackoverflow.com/questions/9110487/undefined-reference-to-a-static-member

File: bolt/include/bolt/Core/DebugData.h [hunk 21]
```diff
@@ -344,14 +344,40 @@ class DebugAddrWriter {
...
   /// Returns False if .debug_addr section was created..
-  bool isInitialized() const { return !AddressMaps.empty(); }
+  bool isInitialized() { return Buffer->size() > 0; }
+
+  /// Updates address base with the given Offset.
+  virtual void updateAddrBase(DIEBuilder &DIEBlder, DWARFUnit &CU,
+                              const uint64_t Offset);
+
+  /// Appends an AddressSectionBuffer to the address writer buffer for the given
+  /// CU.
```
[the remark this answers is not in the document]
Comment (sayhaan) [reply bolt/include/bolt/Core/DebugData.h:366]: The address writer is per CU/DWO CU, right? So wouldn't the CU part still apply?

File: bolt/include/bolt/Core/DebugData.h [hunk 22]
```diff
@@ -344,14 +344,40 @@ class DebugAddrWriter {
...
+  void appendToAddressBuffer(const AddressSectionBuffer &Buffer) {
+    *AddressStream << Buffer;
+  }
+
+  /// Sets AddressByteSize for the CU.
+  void setAddressByteSize(const uint8_t AddressByteSize) {
+    this->AddressByteSize = AddressByteSize;
+  }
+
+  /// Sets AddrOffsetSectionBase for the CU.
```
[the remark this answers is not in the document]
Comment (sayhaan) [reply bolt/include/bolt/Core/DebugData.h:376]: Same for this one, since the AddrOffsetSectionBase is for the AddressWriter and we have one AddressWriter for each CU/DWO CU wouldn't the CU part still apply?

File: bolt/include/bolt/Core/DebugData.h [hunk 23]
```diff
@@ -601,7 +631,7 @@ class DebugLoclistWriter : public DebugLocWriter {
     }
   }
 
-  static void setAddressWriter(DebugAddrWriter *AddrW) { AddrWriter = AddrW; }
+  void setAddressWriter(DebugAddrWriter *AddrW) { AddrWriter = AddrW; }
```
[the remark this answers is not in the document]
Comment (sayhaan) [reply bolt/include/bolt/Core/DebugData.h:634]: Yes, line 636 in DWARFRewriter.cpp

File: bolt/lib/Core/DebugData.cpp (hunk 20 above)
[the remark this answers is not in the document]
Comment (sayhaan) [reply bolt/lib/Core/DebugData.cpp:491]: This is the static one, wouldn't we need to define it in the .cpp file since it's static?

https://stackoverflow.com/questions/35565971/undefined-reference-to-declared-c-static-member-variable

File: bolt/include/bolt/Core/DebugData.h (hunk 23 above)
[the remark this answers is not in the document]
Comment (ayermolo) [reply bolt/include/bolt/Core/DebugData.h:634]: That's for DebugRangeListsSectionWriter::setAddressWriter(
          AddressWriterIterator.get());

File: bolt/lib/Core/DebugData.cpp (hunk 20 above)
[the remark this answers is not in the document]
Comment (ayermolo) [reply bolt/lib/Core/DebugData.cpp:491]: Ah right missed DebugAddrWriter::

File: bolt/include/bolt/Core/DebugData.h (hunk 21 above)
[the remark this answers is not in the document]
Comment (ayermolo) [reply bolt/include/bolt/Core/DebugData.h:366]: "for the given CU" to me implies it's passed somehow.

File: bolt/include/bolt/Core/DebugData.h (hunk 22 above)
[the remark this answers is not in the document]
Comment (ayermolo) [reply bolt/include/bolt/Core/DebugData.h:376]: Comment implies we are doing something for the CU. Maybe rephrase with how it's actually used?

File: bolt/include/bolt/Core/DebugData.h (hunk 23 above)
[the remark this answers is not in the document]
Comment (sayhaan) [reply bolt/include/bolt/Core/DebugData.h:634]: I see, misread the scope. Will remove

File: bolt/lib/Rewrite/DWARFRewriter.cpp [hunk 24]
```diff
@@ -613,27 +613,33 @@ void DWARFRewriter::updateDebugInfo() {
...
     std::lock_guard<std::mutex> Lock(AccessMutex);
     const uint16_t DwarfVersion = CU.getVersion();
     if (DwarfVersion >= 5) {
-      LocListWritersByCU[CUIndex] =
-          std::make_unique<DebugLoclistWriter>(CU, DwarfVersion, false);
+      auto AddrW = std::make_unique<DebugAddrWriterDwarf5>(&BC);
+      std::unique_ptr<DebugAddrWriter> &AddressWriterIterator =
+          AddressWritersByCU.insert({CU.getOffset(), std::move(AddrW)})
+              .first->second;
+      DebugRangeListsSectionWriter::setAddressWriter(
```
Comment (ayermolo) [inline bolt/lib/Rewrite/DWARFRewriter.cpp:636]: This seems off. Inside DebugRangeListsSectionWriter is still static. Probably happens to work because it's all processed serially right now.

File: bolt/lib/Core/DebugData.cpp [hunk 25]
```diff
@@ -395,6 +393,13 @@ DebugAddrWriter::DebugAddrWriter(BinaryContext *BC) : BC(BC) {
   AddressStream = std::make_unique<raw_svector_ostream>(*Buffer);
 }
 
+DebugAddrWriter::DebugAddrWriter(BinaryContext *BC,
+                                 const uint8_t AddressByteSize)
+    : BC(BC), AddressByteSize(AddressByteSize) {
+  Buffer = std::make_unique<AddressSectionBuffer>();
```
Comment (ayermolo) [inline bolt/lib/Core/DebugData.cpp:399]: Remove redundant code, either call this constructor from :DebugAddrWriter(BinaryContext *BC) with some invalid AddressByteSize, or factor out to private "init" function.

File: bolt/lib/Rewrite/DWARFRewriter.cpp [hunk 26]
```diff
@@ -698,7 +712,8 @@ void DWARFRewriter::updateDebugInfo() {
       DWODIEBuilder.updateDWONameCompDirForTypes(DWOStrOffstsWriter,
                                                  DWOStrWriter, **SplitCU,
                                                  DwarfOutputPath, DWOName);
-      DebugLoclistWriter DebugLocDWoWriter(*Unit, Unit->getVersion(), true);
+      DebugLoclistWriter DebugLocDWoWriter(*Unit, Unit->getVersion(), true,
+                                           *AddressWriter);
       DebugRangesSectionWriter *TempRangesSectionWriter = RangesSectionWriter;
       if (Unit->getVersion() >= 5) {
         TempRangesSectionWriter = RangeListsWritersByCU[*DWOId].get();
```
Comment (ayermolo) [inline bolt/lib/Rewrite/DWARFRewriter.cpp:719]: You are not seem to be setting AddressWriter like you do for 
RangeListsSectionWriter->setAddressWriter(AddressWriter);
How is it working?

Comment (sayhaan) [reply bolt/lib/Rewrite/DWARFRewriter.cpp:719]: It's set for RangeListWritersByCU when RangeListWritersByCU is populated, and TempRangesSectionWriter accesses RangeListsWritersByCU. We don't need to populate it in both places as far as I can tell.

Comment (ayermolo) [reply bolt/lib/Rewrite/DWARFRewriter.cpp:719]: Ah I see line 654. Missed it. cool.

File: bolt/lib/Rewrite/DWARFRewriter.cpp [hunk 27]
```diff
@@ -613,48 +613,60 @@ void DWARFRewriter::updateDebugInfo() {
...
                "RangeLists writer for DWO unit already exists.");
         auto RangeListsSectionWriter =
             std::make_unique<DebugRangeListsSectionWriter>();
         RangeListsSectionWriter->initSection(CU);
-        RangeListsWritersByCU[*DWOId] = std::move(RangeListsSectionWriter);
+        DebugRangeListsSectionWriter *RangeListSectionWriter =
+            RangeListsWritersByCU
+                .insert({*DWOId, std::move(RangeListsSectionWriter)})
+                .first->second.get();
+        RangeListSectionWriter->setAddressWriter(AddressWriter);
```
Comment (ayermolo) [inline bolt/lib/Rewrite/DWARFRewriter.cpp:654]: Minor nit. Move above the insert. That way you don't need to do the whole .first->second to store in temp.

File: bolt/lib/Rewrite/DWARFRewriter.cpp [hunk 28]
```diff
@@ -613,48 +613,56 @@ void DWARFRewriter::updateDebugInfo() {
   }
 
   if (BC.isDWARF5Used()) {
-    AddrWriter = std::make_unique<DebugAddrWriterDwarf5>(&BC);
+    FinalAddrWriter = std::make_unique<DebugAddrWriterDwarf5>(&BC);
```
Comment (ayermolo) [inline bolt/lib/Rewrite/DWARFRewriter.cpp:616]: Since we are now passing instances of AddressWriter to update functions let's make FinalAddrWriter local to this function and pass it to finalize functions where it's used.
This way when this function exits we free up memory automatically.

File: bolt/lib/Core/DebugData.cpp [hunk 29]
```diff
@@ -504,55 +502,44 @@ void DebugAddrWriterDwarf5::update(DIEBuilder &DIEBlder, DWARFUnit &CU) {
...
     auto Iter = UnmodifiedAddressOffsets.find(Offset);
-    if (Iter != UnmodifiedAddressOffsets.end()) {
-      updateAddressBase(DIEBlder, *this, CU, Iter->getSecond());
-      return;
-    }
-    UnmodifiedAddressOffsets[Offset] = Buffer->size() + HeaderSize;
-    if (Error Err = AddrTable.extract(AddrData, &Offset, 5, AddrSize,
+    if (Iter != UnmodifiedAddressOffsets.end())
+      return Iter->second;
+    UnmodifiedAddressOffsets[Offset] = BufferSize;
```
[the remark this answers is not in the document]
Comment (sayhaan) [reply bolt/lib/Core/DebugData.cpp:516]: We set UnmodifiedAddressOffsets[Offset] and then return it if we find the Offset in the map later. When we return we call updateAddrBase which adds the HeaderSize, so it's not necessary to add it here.

**Comment by kevinfrei** [comment]
Looks good!

**Review by aaupov — APPROVED** [review]
LGTM
